File: processMain.py
from twisted.internet.protocol import Factory, Protocol
from twisted.internet.endpoints import TCP4ServerEndpoint
from twisted.internet import reactor

from enum import Enum

class ProcessState(Enum):
    DETECT_INIT = 0,
    DETECT_START = 1,
    SENSOR_TRIGGER = 2,
    GPS_TRIGGER = 3,
    CAMERA_TRIGGER = 4,
    ACCIDENT_REPORT = 5


class TCPServer(Protocol):

    def __init__(self, currentState, sensorInitStatus):
        self.currentState = currentState
        self.sensorInitStatus = sensorInitStatus

    def dataReceived(self, data: bytes):
        receivedData = data.decode()

        if self.currentState == ProcessState.DETECT_INIT:
            print("State: Detection init")
            if "SENSOR" in receivedData:
                self.sensorInitStatus[0] = 1
            elif "CAMERA" in receivedData:
                self.sensorInitStatus[1] = 1

            print(self.sensorInitStatus)

            if self.sensorInitStatus[0] == 1 and self.sensorInitStatus[1] == 1:
                self.currentState = ProcessState.DETECT_START

        elif self.currentState == ProcessState.DETECT_START:
            print("State: Detection start")
            if "SENSOR" in receivedData and "TRUE" in receivedData:
                self.currentState = ProcessState.SENSOR_TRIGGER
                
        elif self.currentState == ProcessState.SENSOR_TRIGGER:
            print("State: Sensor triggered")
        elif self.currentState == ProcessState.GPS_TRIGGER:
            print("State: GPS Triggered")
        elif self.currentState == ProcessState.CAMERA_TRIGGER:
            print("State: Camera Triggered")
        elif self.currentState == ProcessState.ACCIDENT_REPORT:
            print("State: Accident Reporting")
        else:
            print("Invalid State")

class TCPServerFactory(Factory):

    def __init__(self):
        self.state = ProcessState.DETECT_INIT
        self.sensorInitStatus = [0, 0]

    def buildProtocol(self, addr):
        return TCPServer(self.state, self.sensorInitStatus)

endpoint = TCP4ServerEndpoint(reactor, 8080)
endpoint.listen(TCPServerFactory())
reactor.run()

File: processMain.js
const net = require("net");

const HOST = "127.0.0.1";
const PORT = 8080;

const STATE = {
    DETECT_INIT: 0,
    DETECT_START: 1,
    SENSOR_TRIGGER: 2,
    GPS_TRIGGER: 3,
    CAMERA_TRIGGER: 4,
    ACCIDENT_REPORT: 5
};

let currentState = STATE.DETECT_INIT;
let sensorInitStatus = [0, 0, 0];

let iothatMessageToSend = "";
let iothatIsMessageAvailable = false;

let accidentData = {
    sensor: "",
    location: "",
    camera: ""
}

net.createServer(socket => {

    socket.on("connect", () => {
        console.log(`Connected to ${socket.remoteAddress}`);
    });

    socket.on("data", data => {
        //TODO: Write data processing stuff
        let receivedData = data.toString();
        console.log(receivedData);
        switch (currentState) {
            case STATE.DETECT_INIT:
                console.log("Current State: Detect Init");
                if (receivedData.indexOf("SENSOR") > -1) {
                    sensorInitStatus[0] = 1;
                } else if (receivedData.indexOf("CAMERA") > -1) {
                    sensorInitStatus[1] = 1;
                } else if (receivedData.indexOf("GPS") > -1) {
                    sensorInitStatus[2] = 1;
                }

                let nextState = true;

                sensorInitStatus.forEach(value => {
                    if (value === 0) {
                        nextState = false;
                    }
                });

                if (nextState === true) {
                    currentState = STATE.DETECT_START;
                }

                break;

            case STATE.DETECT_START:
                console.log("Current State: Detection start");
                if ((receivedData.indexOf("SENSOR") > -1) && (receivedData.indexOf("TRUE") > -1)) {
                    currentState = STATE.SENSOR_TRIGGER;
                    accidentData.sensor = receivedData;
                }
                break;

            case STATE.SENSOR_TRIGGER:
                console.log("Current State: Sensor triggered");
                if ((receivedData.indexOf("GPS") > -1) && (receivedData.indexOf("TRUE") > -1)) {
                    currentState = STATE.GPS_TRIGGER;
                    accidentData.location = receivedData;
                }
                break;

            case STATE.GPS_TRIGGER:
                console.log("Current State: GPS triggered");
                if ((receivedData.indexOf("CAMERA") > -1) && (receivedData.indexOf("TRUE") > -1)) {
                    currentState = STATE.CAMERA_TRIGGER;
                    accidentData.camera = receivedData;
                }
                break;

            case STATE.CAMERA_TRIGGER:
                console.log("Current State: Camera Triggered");

                iothatMessageToSend = `AD -> ${accidentData.sensor}|${accidentData.location}|${accidentData.camera}`;
                iothatIsMessageAvailable = true;

                currentState = STATE.ACCIDENT_REPORT;
                break;

            case STATE.ACCIDENT_REPORT:
                break;

            default:
                break;
        }

        if (receivedData.indexOf("GPS") > -1) {
            if (iothatIsMessageAvailable === true) {
                socket.write(iothatMessageToSend);
                iothatIsMessageAvailable = false;
                iothatMessageToSend = "";
            } else {
                socket.write("EMPTY");
            }
        }
    });

    socket.on("close", () => {
        console.log(`Connection closed: ${socket.remoteAddress}`);
    });

    socket.on("error", error => {
        console.log(`Error at ${socket.remoteAddress}`);
    });

}).listen(PORT, HOST);

console.log(`Raspi Process running at ${HOST}:${PORT}`);
